src/main/java/com/zaxxer/hikari/pool/PoolBase.java

```diff
@@ -145,39 +146,58 @@ void quietlyCloseConnection(final Connection connection, final String closureRea
       }
    }
 
-   boolean isConnectionDead(final Connection connection)
-   {
+   private boolean doIsConnectionDead(final Connection connection) throws SQLException {
+      setNetworkTimeout(connection, validationTimeout);
       try {
-         setNetworkTimeout(connection, validationTimeout);
-         try {
-            final var validationSeconds = (int) Math.max(1000L, validationTimeout) / 1000;
+         final var validationSeconds = (int) Math.max(1000L, validationTimeout) / 1000;
 
-            if (isUseJdbc4Validation) {
-               return !connection.isValid(validationSeconds);
-            }
-
-            try (var statement = connection.createStatement()) {
-               if (isNetworkTimeoutSupported != TRUE) {
-                  setQueryTimeout(statement, validationSeconds);
-               }
+         if (isUseJdbc4Validation) {
+            return !connection.isValid(validationSeconds);
+         }
 
-               statement.execute(config.getConnectionTestQuery());
+         try (var statement = connection.createStatement()) {
+            if (isNetworkTimeoutSupported != TRUE) {
+               setQueryTimeout(statement, validationSeconds);
             }
+
+            statement.execute(config.getConnectionTestQuery());
          }
-         finally {
-            setNetworkTimeout(connection, networkTimeout);
+      }
+      finally {
+         setNetworkTimeout(connection, networkTimeout);
 
-            if (isIsolateInternalQueries && !isAutoCommit) {
-               connection.rollback();
-            }
+         if (isIsolateInternalQueries && !isAutoCommit) {
+            connection.rollback();
          }
+      }
+      return false;
+   }
 
-         return false;
+   private void connectionDeadException(final Connection connection, Exception e) {
+      lastConnectionFailure.set(e);
+      logger.warn("{} - Failed to validate connection {} ({}). Possibly consider using a shorter maxLifetime value.",
+         poolName, connection, e.getMessage());
+   }
+
+   boolean isConnectionDead(final Connection connection)
```

Comment:
The major change is here.

In case of exception let the `exceptionOverride` decide if the connection is good or not, and give it another try since it can be an exception of recovery (e.g. failover switch)